# Project Description

The coronavirus took the entire world by surprise, changing everyone's daily routine. City dwellers no longer spent their free time outside, going to cafes and malls; more people were at home, reading books. That attracted the attention of startups that rushed to develop new apps for book lovers. 

I have been given a database of one of the services competing in this market. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition for a new product. 

# Description of the data

**books**

Contains data on books:

-  book_id
-  author_id
-  title
-  num_page — number of pages
-  publication_date
-  publisher_id

**authors**

Contains data on authors:

-  author_id
-  author

**publishers**

Contains data on publishers:

-  publisher_id 
-  publisher

**ratings**

Contains data on user ratings:

-  rating_id
-  book_id
-  username — the name of the user who rated the book
-  rating

**reviews**

Contains data on customer reviews:

-  review_id
-  book_id
-  username — the name of the user who reviewed the book
-  text — the text of the review

# Project Goal

My project goal is the following:

- Find the number of books released after January 1, 2000.
- Find the number of user reviews and the average rating for each book.
- Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).
- Identify the author with the highest average book rating (look only at books with at least 50 ratings).
- Find the average number of text reviews among users who rated more than 50 books.

**Importing libraries**

In [1]:
import pandas as pd
from sqlalchemy import create_engine

**Creating data base connection:**

In [2]:
db_config = {'user': 'praktikum_student',         # user name
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

**Downloading the "books" database:**

In [3]:
query = ''' 
SELECT * FROM books

        '''
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In the "books" database, there are 1000 rows and 6 columns. Columns name are "book_id", "author_id", "title", "num_pages", "publication_date" and "publisher_id".

**Downloading the "author" database:**

In [4]:
query = ''' 
SELECT * FROM authors

        '''
pd.io.sql.read_sql(query, con = engine)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

In the "author" database, there are 636 rows and 2 columns. Columns name are "authod_id" and "author".

**Downloading the "publishers" database:**

In [5]:
query = ''' 
SELECT * FROM publishers

        '''
pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

In the "publishers" database, there are 340 rows and 2 columns. Columns are "publisher_id" and "publisher".

**Downloading the "ratings" database:**

In [6]:
query = ''' 
SELECT * FROM ratings

        '''
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In the "ratings" database, there are 6456 rows and 4 columns. Columns name are "rating_id", "book_id", "username" and "rating".

**Downloading the "reviews" database:**

In [7]:
query = ''' 
SELECT * FROM reviews

        '''
pd.io.sql.read_sql(query, con = engine)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

In the reviews database, there are 2793 rows and 4 columns. Columns name are "review_id", "book_id", "username" and "text".

**Finding the number of books released after January 1, 2000:**

In [8]:
query = ''' 
SELECT
    COUNT(*)
FROM
    books
WHERE
    publication_date > '2000-01-01';

        '''
pd.io.sql.read_sql(query, con = engine)

count
0    819

The number of released  books are 819 after January 1, 2000.

**Finding the number of user reviews and the average rating for each book:**

In [9]:
query = ''' 
SELECT
    books.title as book_title,
    COUNT(*) AS review_num,
    AVG(rating) as mean_rating
FROM
    books
    INNER JOIN ratings ON ratings.book_id = books.book_id
    INNER JOIN reviews ON reviews.book_id = books.book_id
GROUP BY 
    book_title
ORDER BY 
    review_num DESC;
    '''
pd.io.sql.read_sql(query, con = engine)

book_title  review_num  \
0                              Twilight (Twilight  #1)        1120   
1                  The Hobbit  or There and Back Again         528   
2                               The Catcher in the Rye         516   
3    Harry Potter and the Prisoner of Azkaban (Harr...         492   
4    Harry Potter and the Chamber of Secrets (Harry...         480   
..                                                 ...         ...   
988  The Adventures of Tom Sawyer and Adventures of...           2   
989                                  History of Beauty           2   
990                                         Lysistrata           2   
991                   Executive Orders (Jack Ryan  #8)           2   
992  Dragons of Winter Night (Dragonlance: Chronicl...           2   

     mean_rating  
0       3.662500  
1       4.125000  
2       3.825581  
3       4.414634  
4       4.287500  
..           ...  
988     5.000000  
989     2.500000  
990     4.000000  
991     3.500000  
992     4.000000  

[993 rows x 3 columns]

book_title "Twilight" has the 1120 reviews and mean rating is 3.67.

**Identify the publisher that has released the greatest number of books with more than 50 pages:**

In [10]:
query = ''' 
SELECT
    books.publisher_id,
    COUNT(books.publisher_id),
    publishers.publisher AS publisher
FROM
    books
    INNER JOIN publishers ON publishers.publisher_id = books.publisher_id
WHERE
    num_pages > 50
GROUP BY
    books.publisher_id,
    publisher
ORDER BY 
    count DESC
Limit 1;
    

    
        '''
pd.io.sql.read_sql(query, con = engine)

publisher_id  count      publisher
0           212     42  Penguin Books

"Penguin Books" has released 42 books with more than 50 pages.

**Identifying the author with the highest average book rating (look only at books with at least 50 ratings)**

In [11]:
query = """
SELECT
    authors.author as author,
    AVG(ratings.rating) as average_rating
FROM
    authors
    INNER JOIN books on books.author_id = authors.author_id
    INNER JOIN ratings on ratings.book_id = books.book_id
GROUP BY
    author
HAVING
    COUNT(ratings.rating) > 50
ORDER BY
    average_rating DESC;
 """
pd.io.sql.read_sql(query, con = engine)

author  average_rating
0                          J.K. Rowling/Mary GrandPré        4.288462
1                                     Agatha Christie        4.283019
2                   Markus Zusak/Cao Xuân Việt Khương        4.264151
3                                      J.R.R. Tolkien        4.240964
4                            Roald Dahl/Quentin Blake        4.209677
5                                   Louisa May Alcott        4.203704
6                                        Rick Riordan        4.130952
7                                       Arthur Golden        4.107143
8                                        Stephen King        4.009434
9                                        John Grisham        3.971429
10                                    William Golding        3.901408
11                                    Nicholas Sparks        3.882883
12                                       Jodi Picoult        3.881579
13                                    Sophie Kinsella        3.877193
14                                    James Patterson        3.859375
15                                      J.D. Salinger        3.846939
16           Paulo Coelho/Alan R. Clarke/Özdemir İnce        3.789474
17  William Shakespeare/Paul Werstine/Barbara A. M...        3.787879
18                                          Dan Brown        3.741259
19                                         Lois Lowry        3.738462
20            George Orwell/Boris Grabnar/Peter Škerl        3.729730
21                                    Stephenie Meyer        3.662500
22                                     John Steinbeck        3.643836

"J.K. Rowling/Mary GrandPré" has the highest average book rating.

**Find the average number of text reviews among users who rated more than 50 books:**

In [12]:
query = """ 

SELECT 
    CAST(AVG(rev.count) AS int)
FROM 
    (SELECT 
        username , COUNT(rating_id)
    FROM 
        ratings 
    GROUP BY
        username)  as t, 
        (SELECT
             username, COUNT(username)
        FROM
             reviews
        GROUP BY
             username) as rev
WHERE 
(rev.username = t.username) and t.count>50
    
"""
pd.io.sql.read_sql(query, con = engine)

avg
0   24

 24 is the average number of text reviews among users who rated more than 50 books.

# Overall Conclusion

-  In the "books" database, there are 1000 rows and 6 columns. Columns name are "book_id", "author_id", "title", "num_pages", "publication_date" and "publisher_id".
-  In the "author" database, there are 636 rows and 2 columns. Columns name are "authod_id" and "author".
-  In the "publishers" database, there are 340 rows and 2 columns. Columns are "publisher_id" and "publisher".
-  In the "ratings" database, there are 6456 rows and 4 columns. Columns name are "rating_id", "book_id", "username" and "rating".
-  In the reviews database, there are 2793 rows and 4 columns. Columns name are "review_id", "book_id", "username" and "text".
-  The number of released  books are 819 after January 1, 2000.
-  book_title "Twilight" has the 1120 reviews and mean rating is 3.67.
-  "Penguin Books" has released 42 books with more than 50 pages.
-  "J.K. Rowling/Mary GrandPré" has the highest average book rating.
-  24 is the average number of text reviews among users who rated more than 50 books.